Formula/ksync.rb

```diff
@@ -0,0 +1,36 @@
+class Ksync < Formula
+  desc "Sync files between your local system and a kubernetes cluster"
+  homepage "https://ksync.github.io/ksync/"
+  url "https://github.com/ksync/ksync.git",
+    :tag      => "0.4.5",
+    :revision => "9f40bf134329814a57e1a58d73b84761a2b06c73"
+  sha256 "bd47c5b5418bfc6d2f08abe855a52d4189f8bf380747cdd4e5c1a3ca0a87498a"
+  license "Apache-2.0"
+  depends_on "go" => :build
+
+  def install
+    # Set some stamps and build local CLI binary
+    time = `date +%Y%m%dT%H%M%S`.chomp
+    goversion = "go#{Formula["go"].version}"
+    commit = Utils.safe_popen_read("git", "rev-parse", "--short", "HEAD").chomp
+    system "go", "build", "-ldflags",
+              "-w -X github.com/ksync/ksync/pkg/ksync.GitCommit=#{commit} \
+              -X github.com/ksync/ksync/pkg/ksync.GitTag=#{version} \
+              -X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time} \
+              -X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew \
+              -X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}",
+              *std_go_args,
+              "github.com/ksync/ksync/cmd/ksync"
+  end
+
+  # PLACEHOLDER: Need to find a way of running cleanup before uninstall
+  # def uninstall
+  #   system "ksync", "clean", "--nuke"
+  # end
+
+  test do
+    # Basic build test. Potential for more sophisticated tests in the future
+    # Initialize the local client and see if it completes successfully
+    system "ksync", "--help"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
The problem here is that all commands (including `version`) check for a valid cluster context and connection first. Unless it passes basic connection tests, the command won't work. The only exception is `help` for top level commands.

In our CI tests, we use a local container cluster (using `k3d`) to achieve this, so there's been no need to develop a completely local test. Even the `ksync init --local` command will check for basic health of a connection before completing.

I'm open to other suggestions! (cc @SMillerDev)

Reply:
> If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
You would like to check for the _invalidity_ of a cluster connection? I'm not sure what that would accomplish other than checking the binary is intact (same as `help`) but it can easily be implemented.

Reply:
Yes, checking if the packaging is successful is what we want to test. And more often than not checking `--help` doesn't provide us with that information.

Reply:
In this case I might suggest that it's the same. The only step of relevance here is the build step during the install (if it's not pulling from the bottle, as it should normally be). Since it's a single binary, and command is parsed through it, and would indicate that the build has been successful. Depending on the condition of the environment when `brew test` is run, the expected output could vary.

Reply:
> In this case I might suggest that it's the same.

It's not the same at all. Printing help and version texts are incidental functions, generally implemented as short-circuits that go nowhere near the main code path. A binary that tells you what version it is, but crashes when you try to use it for its intended purpose, is useless. That's why `brew test` needs to exercise some portion of the main functionality, even if it's known to fail.

In software development terms, if the only unit tests you ever wrote were for `help` and `version`, who would trust your code?

Reply:
I understand your point, but what I'm saying is a little different. All functions (and `go test` tests btw) require an active cluster context and healthy connection to execute. The check for these is execute _before_ any of the underlying command is called, meaning that any command (with the exception of `help`) _including_ `version` will first check for this, which as you said, goes nowhere near the "main" code. If the test is to check the validity of the compilation, then running `ksync help` is identical to say running `ksync init` or anything else.

However, since this seems to be a point of contention, I've changed it in the next commit. Note that this hits the same code as any function would (with the exception of help, which actually does more) and stops.